server/bin/pbench-verify-backup-tarballs.py

```diff
@@ -0,0 +1,358 @@
+        nfailed_md5 = sys.maxsize
+    return nfailed_md5
+
+
+def report_failed_md5(backup_obj, tmpdir, report, logger):
+    fail_f = os.path.join(tmpdir, "{}.fail".format(backup_obj.list_name))
+    if os.path.exists(fail_f) and os.path.getsize(fail_f) > 0:
+        try:
+            with open(fail_f) as f:
+                failed_list = f.read()
+        except Exception:
+            # Could not open the file
+            logger.exception(
+                "Could not open the file {}".format(fail_f))
+        else:
+            report.write(
+                "ERROR: in {}: the calculated MD5 of the following entries "
+                "failed to match the stored MD5:\n{}".format(backup_obj.name, failed_list))
+
+
+def compare_with_s3_backup(s3_config_obj, backup_obj, tmpdir, report, logger):
+    if s3_config_obj is None:
+        return Status.FAIL
+
+    # Function to check integrity of results between archive/backup and s3
+    list_s3 = os.path.join(tmpdir, "list.s3")
+    try:
+        with open(list_s3, 'w') as f_list:
+            try:
+                s3_content_list = []
+                kwargs = {'Bucket': s3_config_obj.bucket_name}
+                while True:
+                    resp = s3_config_obj.connector.list_objects(**kwargs)
+                    for obj in resp['Contents']:
+                        f_list.write("{}\n".format(obj['Key']))
+                        md5_returned = obj['ETag'].strip("\"")
+                        s3_content_list.append(Entry(obj['Key'], md5_returned))
+                    try:
+                        kwargs['ContinuationToken'] = resp['NextContinuationToken']
+                    except KeyError:
+                        break
+            except Exception:
+                logger.exception("Something went wrong while listing the objects from S3")
+                return Status.FAIL
+    except Exception:
+        logger.exception(
+            "Could not open the file {}".format(list_s3))
+        return Status.FAIL
+
+    list_backup = os.path.join(tmpdir, backup_obj.list_name)
+    try:
+        with open(list_backup) as f:
+            backup_content = f.readlines()
+    except Exception:
+        logger.exception(
+            "Could not open the file {}".format(list_backup))
+        return Status.FAIL
+    else:
+        backup_content_list = []
+        for content in backup_content:
+            result_name = content.strip("\n")
+            local_result_path = os.path.join(backup_obj.dirname, result_name)
+            local_result_md5 = "{}.md5".format(local_result_path)
+            try:
+                with open(local_result_md5) as k:
+                    md5_local = k.readline().split(" ")[0]
+            except Exception:
+                logger.exception(
+                    "Could not open the file {}".format(local_result_md5))
+                continue
+            else:
+                backup_content_list.append(Entry(result_name, md5_local))
```

Comment:
Lines 146 168: This augmentation of the backup list from a list of names to a list of Entries should be done when the list is first iniaitlized. There should be just one loop: for each tarball create an Entry(name, md5) and add it to the list that is passed around.